8301639: JDI and JDWP specs should clarify potential deadlock issues with method invocation

plummercj opened the pull request:
This PR includes a couple of clarifications of the JDWP and JDI invoke method support. The first is that deadlocks can occur due to a needed "resource" being held. The spec previously just mentioned monitors being held, but this is too specific. The second is to clarify that an invoke on a virtual thread can be more prone to deadlocks than a platform thread may be.

I did a doc build you can look at to view the changes. Links to the relevant sections are below:

https://cr.openjdk.org/~cjplummer/8301639/docs.01/specs/jdwp/jdwp-protocol.html#JDWP_ClassType_InvokeMethod

https://cr.openjdk.org/~cjplummer/8301639/docs.01/specs/jdwp/jdwp-protocol.html#JDWP_ClassType_NewInstance

https://cr.openjdk.org/~cjplummer/8301639/docs.01/specs/jdwp/jdwp-protocol.html#JDWP_InterfaceType_InvokeMethod

https://cr.openjdk.org/~cjplummer/8301639/docs.01/specs/jdwp/jdwp-protocol.html#JDWP_ObjectReference_InvokeMethod

https://cr.openjdk.org/~cjplummer/8301639/docs.01/api/jdk.jdi/com/sun/jdi/ClassType.html#invokeMethod(com.sun.jdi.ThreadReference,com.sun.jdi.Method,java.util.List,int)

https://cr.openjdk.org/~cjplummer/8301639/docs.01/api/jdk.jdi/com/sun/jdi/ClassType.html#newInstance(com.sun.jdi.ThreadReference,com.sun.jdi.Method,java.util.List,int)

https://cr.openjdk.org/~cjplummer/8301639/docs.01/api/jdk.jdi/com/sun/jdi/InterfaceType.html#invokeMethod(com.sun.jdi.ThreadReference,com.sun.jdi.Method,java.util.List,int)

https://cr.openjdk.org/~cjplummer/8301639/docs.01/api/jdk.jdi/com/sun/jdi/ObjectReference.html#invokeMethod(com.sun.jdi.ThreadReference,com.sun.jdi.Method,java.util.List,int)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8316386](https://bugs.openjdk.org/browse/JDK-8316386) to be approved

### Issues
 * [JDK-8301639](https://bugs.openjdk.org/browse/JDK-8301639): JDI and JDWP specs should clarify potential deadlock issues with method invocation (**Bug** - P3)
 * [JDK-8316386](https://bugs.openjdk.org/browse/JDK-8316386): JDI and JDWP specs should clarify potential deadlock issues with method invocation (**CSR**)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15695/head:pull/15695` \
`$ git checkout pull/15695`

Update a local copy of the PR: \
`$ git checkout pull/15695` \
`$ git pull https://git.openjdk.org/jdk.git pull/15695/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15695`

View PR using the GUI difftool: \
`$ git pr show -t 15695`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15695.diff">https://git.openjdk.org/jdk/pull/15695.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15695#issuecomment-1716703810)

plummercj commented:
/label remove core-libs

dholmes-ora commented on src/java.se/share/data/jdwp/jdwp.spec:
```diff
@@ -1168,6 +1168,15 @@ JDWP "Java(tm) Debug Wire Protocol"
+        "are typically user-mode threads scheduled by the Java runtime rather than "
+        "the operating system. As such, there may be more cases with virtual threads "
+        "where the INVOKE_SINGLE_THREADED option may cause a deadlock. "
+        "For example, suspending all threads in the target VM may suspend threads that "
+        "support the timer mechanism for virtual threads, and thus methods such as "
+        "<a href=../../api/java.base/java/lang/Thread.html#sleep(long)>Thread.sleep(long)</a> "
+        "may deadlock."
```
I think this is an implementation detail that you don't really want to document in these unrelated specifications. I'm also unclear why virtual threads are more of a problem here than regular threads, if you have suspended them all - a regular thread won't return from sleep until it is unsuspended. In fact the more I think about this, the less I see why virtual threads need special mention here at all.

dholmes-ora replied:
Of course as soon as I posted that I now realize the issue. There are more things that could "deadlock" virtual threads when all other threads are suspended - such as calling sleep, because it requires a "system" thread to be active. This suggest to me that we need modifications to JDWP/JDI to support effective debugging of virtual threads, by allowing for the notion of "system threads" that by default are not suspended (or else always resumed for an invoke).

AlanBateman replied:
A general point here is that there many cases that are unrelated to virtual threads where suspending all, except one, thread will deadlock/hang if other threads can't make progress. This includes several standard (java.*) APIs.

It is true that there are a couple of additional blocking operations with virtual threads that require I/O or timer threads to make progress in order to avoid a virtual thread being parked or blocked indefinitely. It feels more like something for  "@apiNote" for the invokeMethod/etc. javadoc, not clear to me that it's worth adding as a note in the description of the commands in the JDWP spec.

plummercj replied:
I can change it to `@apiNote`. Since it is not really part of the spec, I can see leaving it out of the JDWP spec. It's mostly important to get this info to JDI users.

AlanBateman replied:
> I can change it to `@apiNote`. Since it is not really part of the spec, I can see leaving it out of the JDWP spec. It's mostly important to get this info to JDI users.

Okay, in which case a CSR is not strictly needed here (I added the "csr" label because of the spec update).

AlanBateman commented:
/csr

AlanBateman commented on src/java.se/share/data/jdwp/jdwp.spec:
```diff
@@ -1147,7 +1147,7 @@ JDWP "Java(tm) Debug Wire Protocol"
         "suspended by an event or by command. "
         "This is done to prevent the deadlocks "
-        "that will occur if any of the threads own monitors "
+        "that will occur if any of the threads own resources, such as monitors, "
```
In passing, shouldn't that be "may occur" rather than the "will occur"?

plummercj replied:
No, because the line below follows with "...that _will_ be needed by the invoked method". If the resource _will_ be needed, then the deadlock _will_ occur.

plummercj commented:
I've updated the javadoc to use `@apiNote`. I updated the jdwp spec to not include the extra language regarding virtual threads, but sill include the "resources" vs "monitors" language. Links to the generated docs have been updated.

dholmes-ora reviewed (approved):
Updates looks good to me.

Thanks

plummercj commented:
CSR has been created: [JDK-8301639](https://bugs.openjdk.org/browse/JDK-8301639)

AlanBateman commented:
> CSR has been created: [JDK-8301639](https://bugs.openjdk.org/browse/JDK-8301639)

I added the "csr" label as it was initially looking like this would involve a spec update. It's now changed to be non-normative and an API note so a CSR isn't strictly needed but okay to have one.

plummercj commented:
> > CSR has been created: [JDK-8301639](https://bugs.openjdk.org/browse/JDK-8301639)
> 
> I added the "csr" label as it was initially looking like this would involve a spec update. It's now changed to be non-normative and an API note so a CSR isn't strictly needed but okay to have one.

In the CR David requested that I still do a CSR to cover the minor jdwp.spec changes. I wasn't intending to have the CSR cover the APi note.

AlanBateman commented:
> In the CR David requested that I still do a CSR to cover the minor jdwp.spec changes. I wasn't intending to have the CSR cover the APi note.

Changes to the JDWP spec or JDI API are tracked by CSR. The changes here just expand rational for why all threads are resumed by default. So not really a change to the spec and not adding or changing something that introduces or changes testable assertions. So a CSR isn't strictly needed here but okay to submit one if you or David want.

plummercj commented:
Thanks for the reviews Alan and David!

/integrate
